### PR TITLE
treewide: big opengl cleanups

### PR DESCRIPTION
**Pull request by K900**
## Description of changes

This is long overdue.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by mweinelt** [comment]
Manual build probably requires updating `nixos/doc/manual/configuration/gpu-accel.chapter.md`.

**Comment by K900** [comment]
Yep, forgot to commit all the renames. Should be there now.

File: nixos/modules/hardware/graphics.nix [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  driversEnv32 = pkgs.buildEnv {
+    name = "graphics-drivers-32bit";
+    paths = [ cfg.package32 ] ++ cfg.extraPackages32;
+  };
+in
+{
+  imports = [
+    (lib.mkRenamedOptionModule [ "services" "xserver" "vaapiDrivers" ] [ "hardware" "opengl" "extraPackages" ])
+    (lib.mkRemovedOptionModule [ "hardware" "opengl" "s3tcSupport" ] "S3TC support is now always enabled in Mesa.")
+    (lib.mkRemovedOptionModule [ "hardware" "opengl" "driSupport"] "The setting can be removed.")
```
Comment (SuperSandro2000) [inline nixos/modules/hardware/graphics.nix:19]: We should probably mention, that this was a no-op in the first place

**Comment by Tungsten842** [comment]
I am not sure that "graphics" is a good name in this case.
Consider that hardware.opengl.extraPackages is currently also used by: [opencl](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/configuration/gpu-accel.chapter.md#intel-sec-gpu-accel-opencl-intel) and level-zero.
Putting compute only APIs in a "graphics" namespace causes unnecessary confusion.
I believe that the previous naming, hardware.drivers made more sense: https://github.com/NixOS/nixpkgs/pull/158079

**Comment by K900** [comment]
`hardware.drivers` is even worse, because it gives you absolutely no idea what "drivers" there are or why you should care. `graphics` is suboptimal, but `accelerators` is way too non-obvious.

**Comment by Atemu** [comment]
The idea would be to have `hardware.drivers.vulkan`, `hardware.drivers.opengl` etc. which would make it quite obvious which drivers there are. Even non-nerdy people use the term "drivers" to refer to this software component. If they typed "vulkan drivers nvidia" into search.nixos.org, I believe they'd find the correct option.

I also take issue with `graphics` though, especially as many signs point to more task-specific accelerator hardware in the future that has nothing to do with graphical processing.   
I'd find enabling drivers for a compression accelerator via `hardware.graphics` equally confusing as configuring Vulkan via `hardware.opengl`. Let's not repeat that mistake.

I agree that `accelerator` is much to technical of a term but I simply know of no better term that is neutral with respect to the task being carried out.

Additionally, uninformed users should never need to be touching these. A typical desktop user for instance should only have to enable a desktop session and/or Steam and those would in turn enable the accelerator APIs typically required for those use-cases.  
Anyone who does need to touch these manually ought to learn some terminology. I think that's a reasonable ask.

I won't block on this as this is already an improvement over the status quo and I've clearly not found the priority to produce code for my preferred options API but please consider my design choices and why I made them.

**Comment by K900** [comment]
Really the problem with separating the options like what you're proposing is that Mesa is all of those at the same time, and Mesa is what most people will be running. Technically, we could separate OpenGL, OpenCL, Vulkan, VAAPI etc lookup paths, since they're symlinks anyway, and just link the right manifests into place, but that feels like a lot of work for questionable gain?

**Comment by Atemu** [comment]
> Really the problem with separating the options like what you're proposing is that Mesa is all of those at the same time

Not quite and that's an issue. It's a geat implementation of most of the most critical and useful acceleration APIs but by far not all of them. I love mesa but I don't think it's a good idea to equate it with hardware acceleration as a whole. There is a world outside of Mesa and it does not support every accel API under the sky and that's an issue if you center your NixOS options around everything being mesa. Think CUDA, ROCm, oneAPI etc. or even just alternative implementations of the APIs mesa supports for which there are valid use-cases.  
This is also again quite graphics-centric thinking as there are far more hardware accelerators than just GPUs but require userspace drivers in much the same way.

> Mesa is what most people will be running

That is true but enforcing that through the design of the NixOS options is the wrong way of achieving that IMO. Ensuring the user has the most sensible drivers installed should be handled through use-case-driven sane defaults.

> Technically, we could separate OpenGL, OpenCL, Vulkan, VAAPI etc lookup paths, since they're symlinks anyway, and just link the right manifests into place, but that feels like a lot of work for questionable gain?

You don't have to do but designing it like that leaves you the option to should that become relevant. In my initial implementation I still threw them into `/run/opengl-driver` but you could totally separate it in the implementation later you wanted to.

Speaking of which, I had a draft with some progress somewhere with some more discussion but I'm having a really hard time finding it. I'll link it here when I find it.

**Comment by Atemu** [comment]
Found it: https://github.com/NixOS/nixpkgs/issues/141803#issuecomment-2010803483

**Comment by K900** [comment]
I do really like your draft. I'll see if I can muster the spoons to actually try implementing it, because it does not seem _too_ difficult now...

**Comment by jonringer** [comment]
I'll defer to @atemu on this matter. I removed most of the references of opengl from packaging concerns, module re-design I like their direction, and looks like the related information was already brought up.

**Comment by K900** [comment]
OK, let's just send this, and I'll try a more ambitious cleanup over the weekend.

**Comment by JohnRTitor** [comment]
Can we backport this? Needed for https://github.com/NixOS/nixos-hardware/pull/997

**Comment by K900** [comment]
Definitiely not.

**Comment by themaxhero** [comment]
> `hardware.drivers` is even worse, because it gives you absolutely no idea what "drivers" there are or why you should care. `graphics` is suboptimal, but `accelerators` is way too non-obvious.

What about hardware.gpu?

**Comment by K900** [comment]
It's not really just GPUs, but anyway, I'll have a PR with a more granular approach to this eventually.

**Comment by JohnRTitor** [comment]
> Definitely not.

Is there a reason not to? Due to nixos-hardware not being versioned, about time we get bug reports from stable users.

**Comment by K900** [comment]
It's an API break. We don't do those on stable releases. nixos-hardware should just use the old options for now, and remove the removed ones.

**Comment by SomeoneSerge** [comment]
Oh, not sure how I missed this. This is great and indeed long overdue, thank you @K900!

Regarding the naming, I think `hardware.graphics` isn't much different from `hardware.opengl` and about as misleading. As you say it's not "just gpus", and it's not "just graphics". I still like `hardware.drivers` (personal preference: [`hardware.impure-drivers`](https://github.com/NixOS/nixpkgs/pull/276558)) as the most precise so far. I am not concerned with how much or little information this name conveys to an uninformed reader on the first encounter. NixOS is in the process of establishing essentially a new concept here: separating pure dependencies/services from those that must be kept impure. As long as the new name (whatever it be) makes sense, is reasonably documented and pointed at with `mkRenamedOptionModule`, it's perfectly reasonable to introduce a new term once in a while.

**Comment by jhvst** [comment]
Why was `setLdLibraryPath` removed?

**Comment by Atemu** [comment]
AFAICT because the last driver relying on this terrible hack was removed (AMDGPU-PRO).

It should receive a `mkRenamedOptionModule` though.

What's your use-case?

**Comment by jhvst** [comment]
I develop Vulkan applications and many of those libraries and programs have hard-coded load paths. The `setLdLibraryPath` helped to set aside some of those problems.

**Comment by K900** [comment]
That should generally not be necessary if you link nixpkgs' libglvnd/vulkan-loader.

**Comment by jhvst** [comment]
Thanks, that answers my question.

**Pull request merged**